src/functional/iterator/processor_interface.py

```diff
@@ -0,0 +1,62 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import enum
+import typing
+
+from functional.iterator import ir as itir
+
+
+class ProcessorKind(enum.Enum):
+    EXECUTOR = 0
+    FORMATTER = 1
```

Comment:
I understand the decision to skip the definition a hierarchy of `Processor`s for now until it is clear that is needed, but the current solution doesn't look much simpler to me. Tagging every processor with a  `ProcessorKind` value is basically the same as defining a Protocol but the definition of their interfaces is not to easy to find (for example, inside the `run_processor()` function in tests). Why do you think is better to define the enum instead of single-method `Protocol`s for each one of the kinds?

Reply:
I have tried a protocol for each. The result is that they are indistinguishable at runtime and you have to add something to distinguish them at runtime anyway. To clarify:

```python
@runtime_checkable
Foo(Protocol):
    def __call__(fencil: FencilDefinition, *args, **kwargs) -> None:
        ...


@runtime_checkable
Bar(Protocol):
    def __call__(fencil: FencilDefinition, *args, **kwargs) -> str:
        ...


# runtime checking does not check any type annotations only existence of the attributes and methods.
isinstance(roundtrip.executor, Foo) == True
isinstance(roundtrip.executor, Bar) == True
```

As you can see `ProcessorKind` is necessary anyway. If the additional protocol seems worth it to you we can add it.

Reply:
A passable alternative is to create wrapper types. I have made that change now and it looks cleaner to me, what do you think?

Reply:
Ok, I see your point. Somehow I hadn't consider this case where all the protocols would use `__call__` for implementation. I had in mind that each protocol defined a different entry point, which would simplify thinks. For example:

```python
@runtime_checkable
class Executor(Protocol):
    def __call__(fencil: FencilDefinition, *args, **kwargs) -> None:
    # or: def execute(fencil: FencilDefinition, *args, **kwargs) -> None:
        ...


@runtime_checkable
class Formatter(Protocol):
    def format_ir(fencil: FencilDefinition, *args, **kwargs) -> str:
        ...
...

```
What do you think about that?

Reply:
In that case I would still want to write the decorators to allow writing a simple function and decorating it. Amounts to almost the same in terms of usability as the current improved state, except for the usage when you start from a `FencilDefinition`. I never minded that you can call `formatter(fencil_ir, *args, **kwargs)`, only that you can call `fencil(*args, backend=formatter, **kwargs)`. The latter violates sane fencil call semantics, the former seems harmless. Your proposal would replace it with `formatter.format_ir(fencil_ir, *args, **kwargs)`.